### DISCUSSION TITLE
Edit buttons in Astroid

### DISCUSSION THREAD

**Discussion by iorbita** [templaza/astroid-framework#327] [Q&A]
Hi 
…need some help please…
Buttons have always been a pain in Joomla templates with Bootstrap (I'm referencing to template's buttons styles and Bootstrap's button styles)
I would like to override some Bootstrap button variables that are in _buttons.scss file, is it possible to override them in the framework? I don't think so... 
If not, which solution do you think is the best?

Thanks,
Lorenzo

![astroid-scss](https://github.com/templaza/astroid-framework/assets/566221/a26fcca9-ff24-4c2e-9fa7-32145709d1da)

**Comment by Chacapamac** [comment]
Yeah, I understand , Bootstrap have tendency to overreach on the design sometimes.
I know you can reach some of the bootstrap settings in the “**Astroid Template Manager**” in the “**Theming**” tab.

![theming](https://github.com/templaza/astroid-framework/assets/1907276/5b965eaf-b46e-4586-abd0-a6ae520353f1)

**You also have a “Sass Overrides”** _(scss)_ **at the end of the “Theming” tab, not sure how it work?**
— You can create variable here and assign some colors, I wonder what happen if you create identical variable as shown in your example.

You can also add an override in the —> media  > templates > site > yourtemplate > scss  folder name —> variable_overrides.scss

![overrides](https://github.com/templaza/astroid-framework/assets/1907276/db5f2083-a8a9-4bc9-97e1-d267e3d5a371)

Not sure if it can help?
or just override it in a custom scss file?

**Comment by iorbita** [comment]
Hi @Chacapamac and thanks for your help,

in _"Theming"_ tab the only section that I think is most appropriate for what I want to do is in _"Sass Overrides"_, _"All Colors"_ and "_Theme Colors"_ sections are too generic.

The rules and variables referring to buttons are situated in **media/astroid/assets/vendor/bootstrap/scss/_buttons.scss** but have also a prefix.

These are the rules I’m interested in:

![astroid1](https://github.com/templaza/astroid-framework/assets/566221/7533215c-f5e9-4f88-9daf-8faab7aab231)

Let's take any example, for example button text colour:

**--#{$prefix}btn-color: #{$body-color};**

How can I override this rule? Is it possible to do this with a variable in _"Sass Overrides"_? I tried in different ways

![astroid3](https://github.com/templaza/astroid-framework/assets/566221/099b16f8-be44-4a91-86b7-1f5f6a2a6c5b)

but it has no impact, clearly what I'm doing is wrong… any idea how to override these rules?

Thanks again,

Lorenzo

**Comment by Chacapamac** [comment]
I think the only way to override bootstrap is to use the **custom scss file** (**_variable_overrides.scss_**) way.

• You can create custom override variables or modified the bootstrap variables via creating a file name “variable_overrides.scss” file in the TEMPLATENAME/scss/custom/variable_overrides.scss folder or in Joomla 4 **media > templates > site > YOURTEMPLATE > scss > _variable_overrides.scss_**

![where](https://github.com/templaza/astroid-framework/assets/1907276/5d772584-027f-4001-bc91-536445441e48)

If you look at that old Joomdev documentation on “How Styling Work in Astroid”  _(we are working on our documentation right now)_
You can see that you can create a first level override of bootstrap by using the “variable_overrides.scss” method
Note: this Joomdev doc is outdated but give good hints.
—> https://github.com/joomdev/Astroid-Framework/wiki/Styling-in-Astroid-Framework

**Below image show you the loading and override priority in Astroid**

![loading](https://github.com/templaza/astroid-framework/assets/1907276/85b71274-c8e5-446c-8138-07f7dbdc556e)

I guess you can even copy the —> **media/astroid/assets/vendor/bootstrap/scss/_buttons.scss** , place it in your “variable_overrides.scss” and modify it from here.

Tell me if this help

**Comment by iorbita** [comment]
Hi and thanks again for your help and for these details, but I think I'm going to drop this issue, I've wasted a lot of time, it seems to be more complicated than I thought, __buttons.scss_ file is quite complex and out of my skills, I use SASS but in a lighter way...

Anyway, rules and variables copied from **media/astroid/assets/vendor/bootstrap/scss/_buttons.scss** to **media /templates/site/MY_TEMPLATE/scss/variable_overrides.scss** are not taken into account, I don't understand why…

Since I needed to edit buttons differently for each template style it would have been nice to be able to edit them in _"Sass Overrides"_ section in each template style, but the task seems to be more difficult.... I hate editing buttons with Bootstrap, it would be nice to find a more friendly solution in Astroid for this task, anyway thanks again for your help ;)

Lorenzo

**Comment by Chacapamac** [comment]
Look at this part of the documentation —> https://docs.astroidframe.work/theming/theming
I think with these settings and a bit of custom scss you can go pretty far in having easy fine tune buttons for different Templates styles.

You can use that html in a page to test Typograpy and Theming.
```
<div class="row">
	<div class="col-sm-12 col-xl-12">
		<h1>H1 Heading</h1>
		<p>Welcome to our wonderful world. We sincerely hope that each and every user entering our website will find exactly what he/she is looking for. With advanced features of activating account and new login widgets, you will definitely have a great experience of using our web page.</p>
		<h2>H2 Heading</h2>
		<p>Welcome to our wonderful world. We sincerely hope that each and every user entering our website will find exactly what he/she is looking for. With advanced features of activating account and new login widgets, you will definitely have a great experience of using our web page.</p>
		<h3>H3 Heading</h3>
		<p>Welcome to our wonderful world. We sincerely hope that each and every user entering our website will find exactly what he/she is looking for. With advanced features of activating account and new login widgets, you will definitely have a great experience of using our web page.</p>
		<h4>H4 Heading</h4>
		<p>Welcome to our wonderful world. We sincerely hope that each and every user entering our website will find exactly what he/she is looking for. With advanced features of activating account and new login widgets, you will definitely have a great experience of using our web page.</p>
		<h5>H5 Heading</h5>
		<p>Welcome to our wonderful world. We sincerely hope that each and every user entering our website will find exactly what he/she is looking for. With advanced features of activating account and new login widgets, you will definitely have a great experience of using our web page.</p>
		<h6>H6 Heading</h6>
		<p>Welcome to our wonderful world. We sincerely hope that each and every user entering our website will find exactly what he/she is looking for. With advanced features of activating account and new login widgets, you will definitely have a great experience of using our web page.</p>
		<hr class="py-2" />
		<h4>Blockquote</h4>
		<blockquote class="bg-light border rounded p-3">
			<p><q>Real estate cannot be lost or stolen, nor can it be carried away. Purchased with common sense, paid for in full, and managed with reasonable care, it is about the safest investment in the world.</q></p>
			<p class="m-0"><cite>Franklin D. Roosevelt</cite></p>
		</blockquote>
	</div>
</div>
<hr class="py-2" />
<h4 class="mb-4">List Style</h4>
<div class="row">
	<div class="col-lg-4">
		<h6>Unordered List</h6>
		<ul>
			<li>Lorem ipsum dolor sit amet consectetur.</li>
			<li>Lorem ipsum dolor sit amet consectetur.</li>
			<li>Lorem ipsum dolor sit amet consectetur.</li>
			<li>Lorem ipsum dolor sit amet consectetur.</li>
			<li>Lorem ipsum dolor sit amet consectetur.</li>
		</ul>
	</div>
	<div class="col-lg-4">
		<h6>Ordered List</h6>
		<ol>
			<li>Lorem ipsum dolor sit amet consectetur.</li>
			<li>Lorem ipsum dolor sit amet consectetur.</li>
			<li>Lorem ipsum dolor sit amet consectetur.</li>
			<li>Lorem ipsum dolor sit amet consectetur.</li>
			<li>Lorem ipsum dolor sit amet consectetur.</li>
		</ol>
	</div>
	<div class="col-lg-4">
		<h6>Icons in a List</h6>
		<ul class="fa-ul">
			<li>Lorem ipsum dolor sit amet consectetur.</li>
			<li>Lorem ipsum dolor sit amet consectetur.</li>
			<li>Lorem ipsum dolor sit amet consectetur.</li>
			<li>Lorem ipsum dolor sit amet consectetur.</li>
			<li>Lorem ipsum dolor sit amet consectetur.</li>
		</ul>
	</div>
</div>
<hr class="py-2" />
<div class="d-flex text-center justify-content-center">
	<div class="col-sm-12">&nbsp;</div>
</div>
<div class="row">
	<div class="col-lg-12">
		<div class="media media-mod-2 align-items-center">&nbsp;</div>
	</div>
</div>
<div class="row pb-5">
	<div class="col-lg-12">
		<div class="media media-mod-2 align-items-center text-right">
			<div class="media-body pr-5 pr-lg-4">&nbsp;</div>
		</div>
	</div>
</div>
<div class="row pb-5">
	<div class="col text-center">
		<button type="button" class="btn btn-primary">Primary</button>
		<button type="button" class="btn btn-secondary">Secondary</button>
		<button type="button" class="btn btn-success">Success</button>
		<button type="button" class="btn btn-danger">Danger</button>
		<button type="button" class="btn btn-warning">Warning</button>
		<button type="button" class="btn btn-info">Info</button>
		<button type="button" class="btn btn-light">Light</button>
		<button type="button" class="btn btn-dark">Dark</button>
	</div>
</div>
<div class="row text-center pb-5">
	<div class="col text-center">
		<h4 class="pb-3">Button tags</h4>
		<a href="#" class="btn btn-primary" role="button">Link</a>
		<button type="submit" class="btn btn-primary">Button</button>
		<input type="button" value="Input" class="btn btn-primary" />
		<input type="submit" value="Submit" class="btn btn-primary" />
		<input type="reset" value="Reset" class="btn btn-primary" />
	</div>
</div>
<div class="row text-center pb-5">
	<div class="col text-center">
		<h4 class="pb-3">Outline buttons</h4>
		<button type="button" class="btn btn-outline-primary">Primary</button>
		<button type="button" class="btn btn-outline-secondary">Secondary</button>
		<button type="button" class="btn btn-outline-success">Success</button>
		<button type="button" class="btn btn-outline-danger">Danger</button>
		<button type="button" class="btn btn-outline-warning">Warning</button>
		<button type="button" class="btn btn-outline-info">Info</button>
		<button type="button" class="btn btn-outline-light">Light</button>
		<button type="button" class="btn btn-outline-dark">Dark</button>
	</div>
</div>
<div class="row text-center pb-5">
	<div class="col text-center">
		<h4 class="pb-3">Sizes</h4>
		<button type="button" class="btn btn-primary btn-sm">Small button</button>
		<button type="button" class="btn btn-secondary btn-sm">Small button</button>
		<div class="clearfix pb-3">&nbsp;</div>
		<button type="button" class="btn btn-primary">Default button</button>
		<button type="button" class="btn btn-secondary">Default button</button>
		<div class="clearfix pb-3">&nbsp;</div>
		<button type="button" class="btn btn-primary btn-lg">Large button</button>
		<button type="button" class="btn btn-secondary btn-lg">Large button</button>
	</div>
</div>
<div class="row text-center pb-5">
	<div class="col text-center">
		<h4 class="pb-3">Button group</h4>
		<div class="btn-group" role="group" aria-label="Basic example">
			<button type="button" class="btn btn-secondary">Left</button>
			<button type="button" class="btn btn-secondary">Middle</button>
			<button type="button" class="btn btn-secondary">Right</button>
		</div>
	</div>
</div>
```

I will test all that soon and add a way to do that more clearly in the documentation.

**Comment by Chacapamac** [comment]
Something interesting here about overriding Bootstrap buttons.
https://stackoverflow.com/questions/28261287/how-to-change-btn-color-in-bootstrap

You can also @import only the bootstrap library you need to customize it.
@import "buttons";
—> https://getbootstrap.com/docs/5.0/customize/optimize/

The ultimate Bootstrap button customization guide —> https://getbootstrap.com/docs/5.2/components/buttons/

I will look at all that, but you right buttons need to be completely adapted to a custom template style.

**Comment by iorbita** [comment]
This is a complex argument as rules are nested within each other, it’s difficult for me to understand if i’s better to override button variables or if it’s better to override button css rules; I also get a lot of confusion with variables used by Bootstrap **var()** and those used by Sass **$**...

I took a look at how Joomla ATUM template was overrided for  buttons: _btn-primary, btn-secondary_ class were created in a stylesheets (in this case __buttons.scss_, look at: **media/templates/administrator/atum/scss/vendor/bootstrap/_buttons.scss** ) using  Bootstrap variables and not Sass variables, in short as I said before, it's complex and there seems to be different approaches to integrate variables...

That's why I really think it would be nice to find a simple solution in Astroid to override important elements like buttons… I’m confident, thanks :)

**Comment by iorbita** [comment]
...after [this tutorial](https://www.youtube.com/watch?v=sFcj7UGUcrE) everything becomes clearer... apparently since version 5.2, Bootstrap has introduced its own variables (so Sass would disappear...?).

So if I understand correctly, Astroid uses Sass variables while Joomla 4 has started to use Bootstrap variables.

**Comment by Chacapamac** [comment]
I will take a look as I will change the official documentation to add tip and trick for Theming. I really want to get all that clearer for people. Customizing buttons is a must in any professional website and it should be easy to do it by using only the template manager.

Without seeing your video, to me, Astroid use Bootstrap and Saas boths and buttons are build with media > Astroid > assets > vendor > bootstrap > scss > _buttons.scss that use bootstrap standard variables —> https://getbootstrap.com/docs/5.2/components/buttons/
Bootstrap use both Sass and their own variables.

I will go deeper there, it’s needed!
